topology/probes/libvirt/libvirt.go

```diff
@@ -0,0 +1,267 @@
+	// The event loop must be registered WITH its poll loop active BEFORE the
+	// connection is opened. Otherwise it just does not work.
+	if err := libvirtgo.EventRegisterDefaultImpl(); err != nil {
+		logging.GetLogger().Errorf("libvirt event handler:  %s", err)
+	}
+
+	go func() {
+		for {
+			if err := libvirtgo.EventRunDefaultImpl(); err != nil {
+				logging.GetLogger().Errorf("libvirt poll loop problem: %s", err)
+			}
+		}
+	}()
+
+	conn, err := libvirtgo.NewConnectReadOnly(probe.uri)
+	if err != nil {
+		logging.GetLogger().Errorf("Failed to create libvirt connect")
+		return
+	}
+	probe.conn = conn
+	callback := func(
+		c *libvirtgo.Connect, d *libvirtgo.Domain,
+		event *libvirtgo.DomainEventLifecycle,
+	) {
+		if event.Event == libvirtgo.DOMAIN_EVENT_STARTED {
+			var interfaces []*Interface
+			probe.queryDomain(d, &interfaces, "")
+			probe.registerInterfaces(interfaces)
+		}
+	}
+	probe.cidLifecycle, err = conn.DomainEventLifecycleRegister(nil, callback)
+	if err != nil {
+		logging.GetLogger().Errorf(
+			"Could not register the lifecycle event handler %s", err)
+	}
+	callbackDeviceAdded := func(
+		c *libvirtgo.Connect, d *libvirtgo.Domain,
+		event *libvirtgo.DomainEventDeviceAdded,
+	) {
+		var interfaces []*Interface
+		probe.queryDomain(d, &interfaces, event.DevAlias)
+		probe.registerInterfaces(interfaces) // 0 or 1 device changed.
+	}
+	probe.cidDevAdded, err = conn.DomainEventDeviceAddedRegister(nil, callbackDeviceAdded)
+	if err != nil {
+		logging.GetLogger().Errorf(
+			"Could not register the device added event handler %s", err)
+	}
+	domains, err := probe.conn.ListAllDomains(0)
+	if err != nil {
+		logging.GetLogger().Error(err)
+		return
+	}
+	var interfaces []*Interface
+	for _, domain := range domains {
+		probe.queryDomain(&domain, &interfaces, "")
+	}
+	probe.registerInterfaces(interfaces)
+}
+
+// Stop stops the probe
+func (probe *Probe) Stop() {
+	probe.graph.RemoveEventListener(probe)
+	if probe.cidLifecycle != -1 {
+		probe.conn.DomainEventDeregister(probe.cidLifecycle)
+		probe.conn.DomainEventDeregister(probe.cidDevAdded)
+	}
+	probe.conn.Close()
+}
+
+// NewProbe creates a libvirt topology probe
+func NewProbe(g *graph.Graph, uri string) *Probe {
+	probe := &Probe{
+		graph:        g,
+		interfaceMap: make(map[string]*Interface),
+		cidLifecycle: -1,
+		uri:          uri,
+	}
+	g.AddEventListener(probe)
+	return probe
+}
+
+// NewProbeFromConfig initializes the probe
+func NewProbeFromConfig(g *graph.Graph) (*Probe, error) {
+	uri := config.GetString("agent.topology.libvirt.url")
+	probe := NewProbe(g, uri)
+	return probe, nil
+}
+
+// OnNodeUpdated event
+func (probe *Probe) OnNodeUpdated(n *graph.Node) {
+}
+
+// OnNodeAdded event
+func (probe *Probe) OnNodeAdded(n *graph.Node) {
+	if typ, _ := n.GetFieldString("Type"); typ != "tun" {
+		return
+	}
+	name, _ := n.GetFieldString("Name")
+	logging.GetLogger().Debugf("Tun interface %s", name)
+	if itf, ok := probe.interfaceMap[name]; ok {
```

Comment:
I think here is an issue: how would this code behave if one interface being attached to multiple vms ? Is it possible ? Not sure about if its possible in hypervisor... but anyway.....
I propose to use my realtime handler which I commited here.
It solves a lot of issues and abstracts away all needed interaction with graph from concrete probe. It will help to reuse this package for future probes... What do you think ? Here is the link to the package which may handle this problem in quite good way
https://github.com/skydive-project/skydive/pull/1424/files#diff-c9c7c72fb06c6b8a0ed7b4c81a3bf3f4